File: dos.py
import numpy as np
import scipy.special
import cmath
import math
import copy
import multiprocessing as mp


import matplotlib
import matplotlib.pyplot as plt

from scipy.sparse.linalg import eigsh
from scipy import sparse
from scipy.signal import windows
# from sparse_dot_mkl import dot_product_mkl


# grid size, timestep and sample size are irrelevant to the shape of the output figure.



t_n = -2.7
t_nn = -0.1
on_site_potential = -0.3


def get_random_state(size):
    state = np.random.random((size, ))
    total = sum([c**2 for c in state])
    return state / math.sqrt(total)

def is_A_site(i, j):
    if (j % 2 == 0 and i % 2 == 1) or (j % 2 == 1 and i % 2 == 0):
        return True
    return False



def flatten_hamiltonian(i, j, horizontal_size, vertical_size):
    rowH = np.zeros((vertical_size, horizontal_size))
    if is_A_site(i, j):
        # nearest neighbors
        rowH[j][(i + 1) % horizontal_size] = -t_n
        rowH[j][(i - 1) % horizontal_size] = -t_n
        rowH[(j - 1) % vertical_size][i] = -t_n

        # # next nearest neighbors
        # rowH[j][(i + 2) % horizontal_size] = -t_nn
        # rowH[(j + 1) % vertical_size][(i + 1) % horizontal_size] = -t_nn
        # rowH[j][(i - 2) % horizontal_size] = -t_nn
        # rowH[(j + 1) % vertical_size][(i - 1) % horizontal_size] = -t_nn
        # rowH[(j - 1) % vertical_size][(i + 1) % horizontal_size] = -t_nn
        # rowH[(j - 1) % vertical_size][(i - 1) % horizontal_size] = -t_nn
        #
        # # on-site potential
        # rowH[j][i] = on_site_potential
    else:
        # nearest neighbors
        rowH[j][(i + 1) % horizontal_size] = -t_n
        rowH[j][(i - 1) % horizontal_size] = -t_n
        rowH[(j + 1) % vertical_size][i] = -t_n

        # # next nearest neighbors
        # rowH[j][(i + 2) % horizontal_size] = -t_nn
        # rowH[(j - 1) % vertical_size][(i + 1) % horizontal_size] = -t_nn
        # rowH[j][(i - 2) % horizontal_size] = -t_nn
        # rowH[(j - 1) % vertical_size][(i - 1) % horizontal_size] = -t_nn
        # rowH[(j + 1) % vertical_size][(i + 1) % horizontal_size] = -t_nn
        # rowH[(j + 1) % vertical_size][(i - 1) % horizontal_size] = -t_nn
        #
        # # on-site potential
        # rowH[j][i] = on_site_potential
    return sparse.csr_matrix(rowH.flatten())


def get_hamiltonian(horizontal_size, vertical_size):
    # pool = mp.Pool(mp.cpu_count())
    # hamiltonian = [pool.apply(flatten_hamiltionian, args=(i, j)) for j in range(vertical_size) for i in range(horizontal_size)]
    hamiltonian = []
    for j in range(vertical_size):
        for i in range(horizontal_size):
            hamiltonian.append(flatten_hamiltonian(i, j, horizontal_size, vertical_size))
        print(j)
    return sparse.vstack(hamiltonian)



# using recursion formula for chebyshev polynomial. x's range is R rather than [-1, 1]
def get_evolution_operator_one_timestep(H, timestep, allowed_error=10**(-13), max_order_of_chebyshev_poly=10000):
    max_eigenvalue = eigsh(H, k=1, which="LA")[0][0]
    min_eigenvalue = eigsh(H, k=1, which="SA")[0][0]
    # max_eigenvalue = 10
    # min_eigenvalue = -10
    print("max_eigenvalue : {}".format(max_eigenvalue))
    print("min_eigenvalue : {}".format(min_eigenvalue))

    operator_size = H.shape[0]
    z = (max_eigenvalue - min_eigenvalue) * timestep / 2
    B = ((H - sparse.identity(operator_size) * (max_eigenvalue + min_eigenvalue) / 2) / (max_eigenvalue - min_eigenvalue)) * (-1j) * 2

    evolution_operator = sparse.csr_matrix((operator_size, operator_size), dtype=np.complex128)
    T_tilde1 = sparse.identity(operator_size)
    T_tilde2 = B
    jv = 1
    i = 1
    while abs(jv) > allowed_error and i <= max_order_of_chebyshev_poly:
        jv = scipy.special.jv(i, z)
        evolution_operator += jv * T_tilde2

        next_T_tilde = B * 2 * T_tilde2 - T_tilde1
        T_tilde1 = T_tilde2
        T_tilde2 = next_T_tilde
        i += 1
        print(i)

    evolution_operator = (evolution_operator * 2 + sparse.identity(operator_size, dtype=np.complex128) * scipy.special.jv(0, z)) * np.exp((max_eigenvalue + min_eigenvalue) * timestep * (-0.5j))
    print("{} : {}".format(i, abs(jv)))
    return sparse.csr_matrix(evolution_operator)



def normalize_state(state):
    total = sum([abs(c)**2 for c in state])
    return state / math.sqrt(total)



def get_correlation(initial_state, evolution_operator, sample_size):
    current_state = copy.deepcopy(initial_state)
    correlations = np.empty((sample_size + 1, ))
    correlations[0] = current_state.dot(current_state)
    for i in range(1, sample_size + 1):
        current_state = normalize_state(evolution_operator.dot(current_state))
        correlations[i] = initial_state.dot(current_state)
        print(i)
    return correlations

def half_hanning(i, N):
    return 0.5 * (1 + np.cos(np.pi * i / N))

def get_dos(correlations, energy_range, timestep):
    # windowed_corr = correlations * windows.hann(correlations.size, sym=False)
    # g = np.fft.fft(windowed_corr) * timestep
    # w = np.fft.fftfreq(windowed_corr.size)*2*np.pi/timestep/abs(t_n)
    # return w, g

    sample_size = correlations.size - 1
    energies = [energy_range / 2 * (-1 + i / sample_size) for i in range(sample_size * 2 + 1)]

    corr_symmetric = np.empty(sample_size * 2 + 1, dtype=np.complex128)
    corr_symmetric[sample_size] = correlations[0]
    for i in range(1, sample_size + 1):
        corr_symmetric[sample_size + i] = half_hanning(i, sample_size) * correlations[i]
        corr_symmetric[sample_size - i] = half_hanning(i, sample_size) * np.conjugate(correlations[i])

    # Fourier transform
    corr_fft = np.fft.fft(corr_symmetric)
    dos = np.empty(sample_size * 2 + 1, dtype=np.complex128)
    for i in range(sample_size * 2 + 1):
        dos[i - sample_size] = corr_fft[i]

    # Normalise
    dos = dos / np.sum(dos)
    return np.array(energies), dos

def draw_dos(horizontal_size, vertical_size, timestep, sample_size):
    random_state = get_random_state(horizontal_size * vertical_size)
    H = get_hamiltonian(horizontal_size, vertical_size)
    evolution_operator = get_evolution_operator_one_timestep(H, timestep)
    corrs = get_correlation(random_state, evolution_operator, sample_size)
    w, g = get_dos(corrs, 6., timestep)
    plt.scatter(w, np.abs(g))
    plt.show()

File: dos.jl
using SparseArrays
using LinearAlgebra
using SpecialFunctions
using DSP
using FFTW
using Plots
using Arpack

const t_n = -2.7

function get_random_state(size::Int64)
    state = rand(size, 1)
    total = sum([abs(c)^2 for c in state])
    return state / sqrt(total)
end

function is_A_site(i::Int64, j::Int64)
    if (j % 2 == 0 && i % 2 == 1) || (j % 2 == 1 && i % 2 == 0)
        return true
    end
    return false
end

function flatten_row_hamiltonian(i::Int64, j::Int64, i_size::Int64, j_size::Int64)
    println(i, ", ", j)
    rowH = zeros(Float64, j_size, i_size)
    if is_A_site(i, j)
        rowH[j, i == i_size ? 1 : i + 1] = -t_n
        rowH[j, i == 1 ? i_size : i - 1] = -t_n
        rowH[j == 1 ? j_size : j - 1, i] = -t_n
    else
        rowH[j, i == i_size ? 1 : i + 1] = -t_n
        rowH[j, i == 1 ? i_size : i - 1] = -t_n
        rowH[j == j_size ? 1 : j + 1, i] = -t_n
    end
    return sparse(reshape(rowH, j_size * i_size, 1))
end

function get_hamiltonian(horizontal_size::Int64, vertical_size::Int64)
    return hcat([flatten_row_hamiltonian(i, j, horizontal_size, vertical_size) for j in 1:vertical_size for i in 1:horizontal_size]...)
end

function get_evolution_operator_onestep_forward(horizontal_size::Int64, vertical_size::Int64, timestep::Float64, allowed_error::Float64=10^(-13), max_order_of_chebyshev_poly::Int64=10000)
    H = get_hamiltonian(horizontal_size, vertical_size)
    #  λ1, ϕ1 = eigs(H, nev=1, which=:LM)
    #  max_eigenvalue = λ1[1]
    max_eigenvalue = 10
    #  λ2, ϕ2 = eigs(H, nev=1, which=:SM)
    #  min_eigenvalue = λ2[1]
    min_eigenvalue = -10

    operator_size = horizontal_size * vertical_size
    z = (max_eigenvalue - min_eigenvalue) * timestep / 2
    B = ((H .- sparse(1.0I, operator_size, operator_size) * (max_eigenvalue + min_eigenvalue) / 2.0) / (max_eigenvalue - min_eigenvalue)) * (-1im) * 2

    evolution_operator = spzeros(operator_size, operator_size)
    T_tilde1 = sparse(1.0I, operator_size, operator_size)
    T_tilde2 = B
    jv = 1
    i = 1
    while abs(jv) > allowed_error && i <= max_order_of_chebyshev_poly
        jv = besselj(i, z)
        evolution_operator = evolution_operator .+ jv * T_tilde2

        next_T_tilde = B * 2 * T_tilde2 .- T_tilde1
        T_tilde1 = T_tilde2
        T_tilde2 = next_T_tilde
        i += 1
        println(i)
    end
    evolution_operator = (evolution_operator * 2 .+ sparse(1.0I, operator_size, operator_size) * besselj(0, z)) * exp((max_eigenvalue + min_eigenvalue) * timestep * (-0.5im))
    return evolution_operator
end

function normalize_state(state)
    total = sum([abs(c)^2 for c in state])
    return state / sqrt(total)
end

function get_correlation(horizontal_size::Int64, vertical_size::Int64, timestep::Float64, sample_size::Int64)
    initial_state = get_random_state(horizontal_size * vertical_size)
    current_state = deepcopy(initial_state)
    evolution_operator = get_evolution_operator_onestep_forward(horizontal_size, vertical_size, timestep)
    correlations= Array{Complex{Float64}}(undef, sample_size)
    for i in 1:sample_size
        current_state = normalize_state(evolution_operator * current_state)
        correlations[i] = dot(initial_state, current_state)
        println(i)
    end
    return correlations
end

function get_dos(correlations::Array{Complex{Float64}, 1}, timestep::Float64)
    #  windowed_corr = correlations .* Windows.hanning(length(correlations), padding=floor(Int64, 0.05 * length(correlations)), zerophase=true)
    windowed_corr = correlations .* Windows.hanning(length(correlations), padding=0, zerophase=true)
    g = fft(windowed_corr) * timestep
    w = [0:(length(correlations)/2-1); (-length(correlations)/2):-1] / length(correlations) * 2 * π / timestep / abs(t_n)
    return w, g
end

function draw_dos(horizontal_size::Int64, vertical_size::Int64, timestep::Float64, sample_size::Int64)
    # horizontal_size = convert(Int64, ARGS[1])
    # vertical_size = convert(Int64, ARGS[2])
    # timestep = convert(Float64, ARGS[3])
    # const
    corrs = get_correlation(horizontal_size, vertical_size, timestep, sample_size)
    w, g = get_dos(corrs, timestep)
    display(plot(w, abs.(g), seriestype = :scatter))
end
